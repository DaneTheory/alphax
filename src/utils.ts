/**
 * data types check util
 * Modified from https://github.com/sindresorhus/is/blob/master/index.js
 */
import * as File from 'vinyl'
import { RenameConfig } from './alphax'

const toString = Object.prototype.toString
const getObjectType = x => toString.call(x).slice(8, -1)
const isOfType = type => x => typeof x === type // eslint-disable-line valid-typeof
const isObjectOfType = type => x => getObjectType(x) === type

export const isBoolean = x => x === true || x === false

export const isSymbol = isOfType('symbol')

export const isInteger = Number.isInteger

export const isSafeInteger = Number.isSafeInteger

export const isString = isOfType('string')

export const isUndefined = isOfType('undefined')

export const isNull = x => x === null

export const isNullOrUndefined = x => isUndefined(x) || isNull(x)

export const isFunction = isOfType('function')

export const isBuffer = Buffer.isBuffer

export const isGeneralizedObject = x => typeof x === 'object'

export const isPlainObject = isObjectOfType('Object')

export const isObject = x => !isNullOrUndefined(x) && (isFunction(x) || isGeneralizedObject(x))

const isValidLength = x => isSafeInteger(x) && x > -1;

export const isArray = Array.isArray

export const isArrayLike = x => !isNullOrUndefined(x) && !isFunction(x) && isValidLength(x.length);

export const isNativePromise = isObjectOfType('Promise')

const hasPromiseAPI = x => isObject(x) && isFunction(x.then) && isFunction(x.catch)

export const isPromise = x => isNativePromise(x) || hasPromiseAPI(x)

export const isIterable = x => !isNullOrUndefined(x) && isFunction(x[Symbol.iterator])

export const isGenerator = x => isIterable(x) && isFunction(x.next) && isFunction(x.throw)

export const isRegExp = isObjectOfType('RegExp')

export function curryFileTransformer(fn) {
  return function (file: File, cb) {
    const res = fn(file)
    if (isPromise(res)) {
      res.then(asyncRes => {
        if (asyncRes === null) {
          cb()
        } else {
          cb(null, file)
        }
      }).catch(error => {
        console.log(error)
      })
    } else {
      if (res === null) {
        cb()
      } else {
        cb(null, file)
      }
    }
  }
}

export function getRenamerByConfig(renameConfig: RenameConfig, context: any): (filepath: string) => string {
  return function (filepath: string) {
    Object.keys(renameConfig).forEach(pattern => {
      let condition = renameConfig[pattern]
      if (/(\?|\=)/.test(condition)) {
        try {
          condition = evaluate(condition, context)
        } catch (error) {
          console.log(error)
        }
      }
      if (isNull(condition)) {
        return
      }
      filepath = filepath.replace(new RegExp(pattern, 'g'), condition)
    })
    return filepath
  }
}

export function evaluate(exp: string | boolean, data: any) {
  /* eslint-disable no-new-func */
  const fn = new Function('data', 'with (data) { return ' + exp + '}')
  try {
    return fn(data)
  } catch (e) {
    throw new Error('Error when evaluating filter condition: ' + exp)
  }
}
