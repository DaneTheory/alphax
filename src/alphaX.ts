import path from 'path'
import fs from 'fs-extra'
import { EventEmitter } from 'events'
import ware from 'ware'
import dest from './dest'
import { isArray, isPromise } from "./utils"

type Middleware = (ctx: AlphaX) => any
type Glob = string[]
type TransformFn = (contents: string) => Promise<string> | string

interface RenameConfig {
  [oldname: string]: string
}

interface filterConfig {
  [pattern: string]: boolean
}

class AlphaX extends EventEmitter {
  private middlewares: Middleware[]
  private patterns: Glob
  private renameConfig: RenameConfig
  private filterConfig: filterConfig
  private dotFiles: boolean
  public meta: any
  public baseDir: string
  public transformFn: TransformFn

  constructor() {
    super()
    this.middlewares = []
    this.meta = {}
  }

  src(patterns: Glob, {
    baseDir = '.',
    rename = {},
    filter = {},
    transformFn,
    dotFiles = true,
  } = {}) {
    this.baseDir = path.resolve(baseDir)
    this.patterns = isArray(patterns) ? patterns : [patterns]
    this.dotFiles = dotFiles
    this.renameConfig = rename
    this.filterConfig = filter
    this.transformFn = transformFn
    return this
  }

  getNewName(name: string): string {
    Object.keys(this.renameConfig).forEach(pattern => {
      name = name.replace(pattern, this.renameConfig[pattern])
    })
    return name
  }

  async transformer(file) {
    // 1. middleware
    await new Promise((resolve, reject) => {
      ware().use(this.middlewares).run(file, this, (err: Error) => {
        if (err) return reject(err)
        resolve()
      })
    })

    // 2. rename
    let oldRelative = file.relative
    let newName = path.join(file.base, this.getNewName(file.relative))
    if (file.path !== newName) {
      file.path = newName
      this.emit('rename', {
        oldname: oldRelative,
        newname: file.relative
      })
    }

    // 3. transform
    if (!file.isDirectory()) {
      const contents = file.contents.toString()
      let result = this.transformFn(contents)
      if (isPromise(result)) {
        result = await result
      }
      file.contents = Buffer.from(result)
    }
  }

  use(middleware: Middleware) {
    this.middlewares.push(middleware)
    return this
  }

  async dest(destPath: string, {
    clean = false
  } = {}) {
    destPath = path.resolve(this.baseDir, destPath)

    if (clean) {
      await fs.remove(destPath)
    }

    if (this.filterConfig) {
      Object.keys(this.filterConfig).forEach(fileName => {
        if (!this.filterConfig[fileName]) {
          this.patterns.push('!./**/' + fileName)
        }
      })
    }

    await dest(
      this.patterns,
      destPath,
      {
        allowEmpty: true,
        transformer: this.transformer.bind(this)
      }
    )
  }
}

export default () => new AlphaX()
