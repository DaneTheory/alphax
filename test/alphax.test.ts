import alphaX from '../src/alphax'
import { SRC_DIR, getDistDir, globDir } from './utils'

describe('alphax', () => {

  test('base', async () => {
    const DIST_DIR = getDistDir('alphax-base')
    const app = alphaX()
    await app
      .src(
        SRC_DIR + '/**',
        {
          filter: {
            'a': true,
            'a/b/**': false,
            'd.js': false
          },
          rename: {
            'a': 'A',
            '.js': '.ts'
          },
          transformFn(content) {
            return `/* Created at ${new Date().toLocaleTimeString()} */` + content
          }
        }
      )
      .dest(DIST_DIR)
    const files = await globDir(DIST_DIR, { baseDir: DIST_DIR })
    expect(files).toMatchSnapshot()
  })

})
