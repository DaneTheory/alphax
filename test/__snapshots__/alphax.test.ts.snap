// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`alphax base 1`] = `
Array [
  "A",
  "A/b.ts",
]
`;
